### PR TITLE
NIFI-12982 Extend test suite of MockProcessSession

### PR DESCRIPTION
**Pull request by EndzeitBegins**
# Summary

[NIFI-12982](https://issues.apache.org/jira/browse/NIFI-12982)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java [hunk 1]
```diff
@@ -45,189 +52,588 @@
...
-        fillBuffer(in, buffer);
+    private final Processor processor = new TestProcessor();
+    private final SharedSessionState sharedState = new SharedSessionState(processor, new AtomicLong(0L));
+    private final MockStateManager stateManager = new MockStateManager(processor);
+    private final MockProcessSession session = new MockProcessSession(sharedState, processor, stateManager);
+
+    private final Processor statefulProcessor = new StatefulTestProcessor();
+    private final SharedSessionState sharedStateOfStatefulProcessor = new SharedSessionState(statefulProcessor, new AtomicLong(0L));
+    private final MockStateManager stateManagerOfStatefulProcessor = new MockStateManager(statefulProcessor);
+    private final MockProcessSession sessionOfStatefulProcessor = new MockProcessSession(sharedStateOfStatefulProcessor, statefulProcessor, stateManagerOfStatefulProcessor);
```
Comment (exceptionfactory) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java:63]: As a general rule, reusing an instance of an object across test method invocations can lead to unexpected results.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for working on additional tests for `MockProcessSession` @EndzeitBegins. Instead of creating `final` instances of various objects, did you consider creating them in a more localized scope? Building on the `Nested` approach, it is possible to have before and after methods that could apply to particular nested groups of tests if needed.

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 1 above)
Comment (EndzeitBegins) [reply nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java:63]: Thank you for your feedback @exceptionfactory. 
I totally agree!

However, as far as I'm aware the default `@TestInstance` lifecycle of JUnit 5 is `PER_METHOD`. 
That is, JUnit 5 is creating a new instance of the `TestMockProcessSession` class for every test method invoked. 
As a result, those objects aren't actually reused between test methods. 

I started out having the "stateful" variants in a single nested scope, but ended up needing them in another nested scope, as far as I remember, which is why I moved them up. 

I'm fine fine with making these field non final and moving those instantiations in a `@BeforeEach` if that's preferred. From my perspective (using mainly Kotlin) it's just more verbose.

Comment (exceptionfactory) [reply nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java:63]: Thanks for the helpful reply @EndzeitBegins, very good point regarding the standard per-method behavior. In light of this, and the fact that there are no Mockito-based mock objects, the current approach makes sense.

**Review by exceptionfactory — APPROVED** [review]
Thanks again for the work on this test class, and for the helpful reply. Based on a further review, this looks good. +1 merging

**Comment by EndzeitBegins** [comment]
You're welcome. Thank you for the review and merge, @exceptionfactory.

**Pull request closed without merging**